### PR TITLE
Add support for Cyclic Learning Rate

### PR DESCRIPTION
**Pull request by mhmoodlan**
Added support for cyclic learning rate as described in: https://arxiv.org/pdf/1506.01186.pdf
Also added functional tests for each policy.

### CODE REVIEW

**Comment by mhmoodlan** [comment]
@Reviewers, is there any other file i should modify?

**Comment by FancyCoder0** [comment]
Hi Mahmoud Aslan(@mhmoodlan),
We are researchers working on identifying redundant development and duplicated pull requests. We have found there is a pull request: https://github.com/tensorflow/tensorflow/pull/20756 which might be a potentially duplicate to this one. We would like to build the link between developers to reduce redundant development. We would really appreciate if you could help us to validate and give us feedback.
Thank you very much for your time!

**Comment by mhmoodlan** [comment]
Hi Luyao Ren(@FancyCoder0),
The pull request you are referring to ([#20756](https://github.com/tensorflow/tensorflow/pull/20756)) is closed, however, they opened another one ([20758](https://github.com/tensorflow/tensorflow/pull/20758)) considering the same idea and they explain in this [comment](https://github.com/tensorflow/tensorflow/pull/20758#issuecomment-405236311) the main difference between their pull request ([20758](https://github.com/tensorflow/tensorflow/pull/20758)) and this one.

**Comment by qlzh727** [comment]
@sguada, not sure whether you are the best person here, can u take a look if you don't mind?

**Comment by futurely** [comment]
It would be good to see some experiment results to show the advantages of the implementation.

https://github.com/tensorflow/tensorflow/pull/20758#issuecomment-413770310

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 74 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 89 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 104 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 119 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @martinwicke, @sguada: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/python/training/learning_rate_decay_test.py [hunk 1]
```diff
@@ -29,6 +29,68 @@
 from tensorflow.python.training import learning_rate_decay
 
 
+class CyclicLearningRateTest(test_util.TensorFlowTestCase):
+  """Functional test for Cyclic Learning Rate"""
+  
+  def np_cyclic_learning_rate(self, step, lr, max_lr, step_size, mode):
+    cycle = math.floor(1. + step / (2. * step_size))
+    x = math.fabs(step / step_size - 2. * cycle + 1.)
+    clr = (max_lr - lr) * max(0., 1. - x)
```
Comment (sguada) [inline tensorflow/python/training/learning_rate_decay_test.py]: Different than https://github.com/bckenstler/CLR
cycle = np.floor(1+step/(2*step_size))
x = np.abs(step/step_size - 2*cycle + 1)
clr = base_lr + (max_lr-base_lr)*np.maximum(0, (1-x))

File: tensorflow/python/training/learning_rate_decay_test.py [hunk 2]
```diff
@@ -29,6 +29,68 @@
...
+    if mode == 'triangular2':
+      clr = clr/(math.pow(2, (cycle - 1)))
+    if mode == 'exp_range':
+      clr = clr * math.pow(.99994, step)
```
Comment (sguada) [inline tensorflow/python/training/learning_rate_decay_test.py]: Define gamma=0.99994
clr= base_lr + (max_lr-base_lr)*np.maximum(0, (1-x))*gamma**(step)

File: tensorflow/python/training/learning_rate_decay_test.py (hunk 1 above)
Comment (mhmoodlan) [reply tensorflow/python/training/learning_rate_decay_test.py]: Thanks @sguada,
Sorry I didn't catch the difference, are you referring to the usage of Numpy instead of python's math module?
For the last statement: "`clr = base_lr + (max_lr-base_lr)*np.maximum(0, (1-x))`", I am delaying the addition of `base_lr` till the end, after handling all possible modes.

**Review by alextp — CHANGES_REQUESTED** [review]
I think this should first go in tensorflow/addons; can you file a pull request there instead?

**Comment by karmel** [comment]
Closing given @alextp 's comment above.
